circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,213 @@
+#pragma once
+
+#include "init.hpp"
+
+#include <aztec3/circuits/abis/public_kernel/public_kernel_inputs_no_previous_kernel.hpp>
+#include <aztec3/circuits/abis/public_kernel/public_kernel_inputs.hpp>
+#include <aztec3/circuits/abis/kernel_circuit_public_inputs.hpp>
+#include <aztec3/circuits/abis/state_read.hpp>
+#include <aztec3/circuits/abis/state_transition.hpp>
+#include <aztec3/circuits/abis/public_data_write.hpp>
+#include <aztec3/utils/dummy_composer.hpp>
+#include <aztec3/utils/array.hpp>
+#include <aztec3/circuits/hash.hpp>
+
+using NT = aztec3::utils::types::NativeTypes;
+using aztec3::circuits::abis::KernelCircuitPublicInputs;
+using aztec3::circuits::abis::PublicDataWrite;
+using aztec3::circuits::abis::StateRead;
+using aztec3::circuits::abis::StateTransition;
+using aztec3::circuits::abis::public_kernel::PublicKernelInputs;
+using aztec3::circuits::abis::public_kernel::PublicKernelInputsNoPreviousKernel;
+using DummyComposer = aztec3::utils::DummyComposer;
+using aztec3::circuits::root_from_sibling_path;
+using aztec3::utils::array_length;
+using aztec3::utils::array_pop;
+using aztec3::utils::array_push;
+using aztec3::utils::push_array_to_array;
+
+namespace aztec3::circuits::kernel::public_kernel {
+
+template <typename NT, size_t SIZE>
+void check_membership(DummyComposer& composer,
+                      typename NT::fr const& value,
+                      typename NT::fr const& index,
+                      std::array<typename NT::fr, SIZE> const& sibling_path,
+                      typename NT::fr const& root)
```

Comment:
I'd maybe add a `std::string const& msg` parameter here, to convey a more bespoke error message if the check fails. E.g. if the check fails because of one of the calls in `validate_state_reads`, we could provide a `msg` from that function of the form:
`const std::string msg = format("validate_state_reads", "; index: ", i, "; leaf_value: ", leaf_value)` (and maybe more debugging info, as you wish). Note: I forget how the `format` function works exactly. There are likely examples in bberg / Aztec Connect. 

Then the eventual error message given by the composer could be the concatenation of `Membership check failed` and the `msg`.